src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt

```diff
@@ -1039,8 +1039,12 @@ class RsTypeInferenceWalker(
 
     private fun inferTryExprType(expr: RsTryExpr): Ty {
         val base = expr.expr.inferType() as? TyAdt ?: return TyUnknown
-        // TODO: make it work with generic `std::ops::Try` trait
-        if (base.item != items.Result && base.item != items.Option) return TyUnknown
+        if (base.item != items.Result && base.item != items.Option) {
+            val (implLookup, knownItems) = expr.implLookupAndKnownItems
+            val tryItem = knownItems.Try ?: return TyUnknown
+            val result = implLookup.select(TraitRef(base, BoundElement(tryItem))).ok() ?: return TyUnknown
+            return result.impl.associatedTypesTransitively.find { it.name == "Ok" }?.declaredType?.substitute(result.subst) ?: TyUnknown
```

Comment:
Let's do `ctx.normalizeAssociatedTypesIn(TyProjection.valueOf(base, tryItem, tryItem.findAssociatedType("Ok")))` instead. This is better in the case of partially inferred types

Reply:
Ok! Thanks, this also looks cleaner.